edb/server/compiler/compiler.py

```diff
@@ -537,6 +539,65 @@ def _compile_ql_script(
 
         return sql.decode(), argmap
 
+    def _compile_query_args(
+            self, 
+            ctx: CompileContext,
+            subtypes: List[Tuple[str, Any, bool]], 
+    ) -> Tuple[bytes, uuid.UUID]:
+        current_tx = ctx.state.current_tx()
+        schema = current_tx.get_schema(self._std_schema)
+        elements = []
+        for name, schema_type, required in subtypes:
+            # todo fmoor: this doesn't handle qualified type names
+            type_name = schema_type.get_name(schema).get_local_name().name
+
+            elem = qlast.ShapeElement(
+                expr=qlast.Path(steps=[qlast.Ptr(
+                    ptr=qlast.ObjectRef(name=f'__arg_{name}'),
+                    direction=s_pointers.PointerDirection.Outbound,
+                )]),
+                compexpr=qlast.TypeCast(
+                    expr=qlast.Parameter(name=name, optional=not required),
```

Comment:
The next puzzle here is that it doesn't matter if `optional` is set to `True` or `False`. In both cases the field cardinality in the resulting descriptor data is `0x41 (ONE)`. It should be `0x6f (AT_MOST_ONE)` when `optional=True`.

Any ideas @elprans?

Reply:
This seems like a bug, let me see.

Reply:
You need to specify optionality on the `TypeCast` node, not `Parameter`.  I think `Parameter.optional` is a vestige that needs removal.